conn.go

```diff
@@ -156,34 +157,86 @@ func (cc *Conn) netlinkConnUnderLock() (*netlink.Conn, netlinkCloser, error) {
-	}
+// receiveSeq returns an iterator of messages to be read from the provided
+// netlink connection filtering out non-nftables messages. It will stop
+// iterating when the buffer is drained or in the case of a fatal error.
+// Non-fatal errors encountered while receiving messages are yielded along with
+// a zero-value message.
+func (cc *Conn) receiveSeq(conn *netlink.Conn) iter.Seq2[netlink.Message, error] {
+	return func(yield func(netlink.Message, error) bool) {
+		if conn == nil {
+			yield(netlink.Message{}, errors.New("netlink conn is not initialized"))
+			return
+		}
 
-	// first receive will be the message that we expect
-	reply, err := nlconn.Receive()
-	if err != nil {
-		return nil, err
-	}
+		for {
+			ready, err := cc.isReadReady(conn)
+			if err != nil {
+				yield(netlink.Message{}, err)
+				return
+			}
 
-	if (sentMsgFlags & netlink.Acknowledge) == 0 {
-		// we did not request an ack
-		return reply, nil
-	}
+			// Since SendMessages is blocking and netlink communication is
+			// synchronous, the kernel has already processed the request and queued
+			// any responses by the time SendMessages returns. Therefore, if
+			// isReadReady returns false on the first call, it means there are no
+			// messages coming at all and we can safely exit.
+			if !ready {
+				break
+			}
 
-	if (sentMsgFlags & netlink.Dump) == netlink.Dump {
-		// sent message has Dump flag set, there will be no acks
-		// https://github.com/torvalds/linux/blob/7e062cda7d90543ac8c7700fc7c5527d0c0f22ad/net/netlink/af_netlink.c#L2387-L2390
-		return reply, nil
+			replies, err := conn.Receive()
+			if err != nil {
+				// Yield the error but continue iterating
+				if !yield(netlink.Message{}, err) {
+					return
+				}
+				continue
+			}
+
+			if len(replies) == 0 && cc.TestDial != nil {
+				// When using a test dial function, we don't always get a reply for each
+				// sent message. Additionally, there is no buffer to poll for more data,
+				// so we stop here.
+				return
+			}
+
+			for _, msg := range replies {
+				// Filter out non-nftables messages.
+				// In practice, this would only be netlink.Error messages.
+				// Those are handled by the netlink library itself and should be
+				// reported as errors by conn.Receive().
+				subsystem := msg.Header.Type >> 8
+				if subsystem != unix.NFNL_SUBSYS_NFTABLES {
+					continue
+				}
+
+				// Stop iteration if yield returns false
+				if !yield(msg, nil) {
+					return
+				}
+			}
+		}
 	}
+}
 
-	// Now we expect an ack
-	if _, err = nlconn.Receive(); err != nil {
-		return nil, err
+// receive will drain the receive buffer of the provided netlink connection
+// and return all received messages, along with the first error encountered,
+// if any.
+func (cc *Conn) receive(conn *netlink.Conn) ([]netlink.Message, error) {
+	var allReplies []netlink.Message
+	var firstError error
+
+	for msg, err := range cc.receiveSeq(conn) {
+		if err != nil && firstError == nil {
```

Comment:
receiveAckAware returns the first receive error, flush joins all receive errors. Perhaps we could join them here too for consistency

Reply:
I’m in favor of consistency, but I wonder if we should go the other way and make Flush also only return the first error? I’m not sure how useful the non-first errors are…